sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/basicOperators.scala

```diff
@@ -40,34 +40,69 @@ case class Project(projectList: Seq[NamedExpression], child: LogicalPlan) extend
  * output of each into a new stream of rows.  This operation is similar to a `flatMap` in functional
  * programming with one important additional feature, which allows the input rows to be joined with
  * their output.
+ * @param generator the generator expression
  * @param join  when true, each output row is implicitly joined with the input tuple that produced
  *              it.
  * @param outer when true, each input row will be output at least once, even if the output of the
  *              given `generator` is empty. `outer` has no effect when `join` is false.
- * @param alias when set, this string is applied to the schema of the output of the transformation
- *              as a qualifier.
+ * @param child Children logical plan node
+ * @param qualifier Qualifier for the attributes of generator(UDTF)
+ * @param attributeNames the column names for the generator(UDTF), will be _c0, _c1 .. _cN if
+ *                       leave as default (empty)
  */
 case class Generate(
     generator: Generator,
     join: Boolean,
     outer: Boolean,
-    alias: Option[String],
-    child: LogicalPlan)
+    child: LogicalPlan,
+    qualifier: Option[String] = None,
+    attributeNames: Seq[String] = Nil)
   extends UnaryNode {
 
-  protected def generatorOutput: Seq[Attribute] = {
-    val output = alias
-      .map(a => generator.output.map(_.withQualifiers(a :: Nil)))
-      .getOrElse(generator.output)
-    if (join && outer) {
-      output.map(_.withNullability(true))
-    } else {
-      output
+  override lazy val resolved: Boolean = {
+    generator.resolved && childrenResolved && attributeNames.length > 0
+  }
+
+  def generatorOutput(): Seq[Attribute] = {
+    if (_generatorOutput == null) {
+      val elementTypes = generator.elementTypes
+
+      val raw = if (attributeNames.size == elementTypes.size) {
+        attributeNames.zip(elementTypes).map {
+          case (n, (t, nullable)) => AttributeReference(n, t, nullable)()
+        }
+      } else {
+        elementTypes.zipWithIndex.map {
+          // keep the default column names as Hive does _c0, _c1, _cN
+          case ((t, nullable), i) => AttributeReference(s"_c$i", t, nullable)()
+        }
+      }
+
+      _generatorOutput = qualifier.map(q => raw.map(_.withQualifiers(q :: Nil))).getOrElse(raw)
     }
+
+    _generatorOutput
   }
 
-  override def output =
-    if (join) child.output ++ generatorOutput else generatorOutput
+  private var _generatorOutput: Seq[Attribute] = null
```

Comment:
I know that this is similar to how I implemented some of this originally, but I think my design was actually a really bad idea.  It would be much better if we could make the `Seq[Attributes]` an argument to the constructor.  That way we know it will be copied correctly when this node or its children are transformed.